### PR TITLE
fix some alerts found with lgtm

### PR DESCRIPTION
**Pull request by jhelie**
Hi,
I just wanted to quickly fix some alerts flagged by lgtm code queries. Most were minor/style but a few could result in a crash.

The standard lgtm deep code analysis highlights dependencies, vulnerabilities and currently flag 226 alerts for scipy (https://lgtm.com/projects/g/scipy/scipy/) - but you can investigate further by writing custom queries (https://lgtm.com/docs/ql/about-ql).

You might also want to try to turn PR integration on to get the chance to investigate new alerts before they find their way into the code base - see https://lgtm.com/docs/lgtm/config/pr-integration.

Hope that helps! and thanks for making scipy what it is :)

### CODE REVIEW

**Review by jaimefrio — APPROVED** [review]

**Comment by jaimefrio** [comment]
Thanks Jean! These changes LGTM, although I am not familiar with some of the modules involved.

**Comment by jhelie** [comment]
You're welcome! Happy to comment on these if some are not clear. I think the most serious one was in `optimize/nonlin.py` and then some exceptions in a few modules that were not raised properly. The reference comparison in `linalg/special_matrices.py` instead of value comparison was interesting too. The rest was mostly useless code (e.g. redundant check, duplicate line, `else` statement after `for` block without break clause...) and straightforward.

File: scipy/sparse/linalg/eigen/arpack/setup.py [hunk 1]
```diff
@@ -16,8 +16,6 @@ def configuration(parent_package='',top_path=None):
     if not lapack_opt:
         raise NotFoundError('no lapack/blas resources found')
 
-    config = Configuration('arpack', parent_package, top_path)
```
Comment (perimosocordiae) [inline scipy/sparse/linalg/eigen/arpack/setup.py]: I'd prefer to get rid of the earlier copy of this line (on line 12) instead.

Comment (jhelie) [inline scipy/sparse/linalg/eigen/arpack/setup.py]: ah, I hesitated for a while. will update!

File: scipy/sparse/linalg/eigen/arpack/setup.py [hunk 2]
```diff
@@ -9,14 +9,12 @@ def configuration(parent_package='',top_path=None):
...
 
-    config = Configuration('arpack',parent_package,top_path)
-
     lapack_opt = get_info('lapack_opt')
 
     if not lapack_opt:
         raise NotFoundError('no lapack/blas resources found')
 
-    config = Configuration('arpack', parent_package, top_path)
+    config = Configuration('arpack',parent_package,top_path)
```
Comment (perimosocordiae) [inline scipy/sparse/linalg/eigen/arpack/setup.py]: With spaces after the commas, please. :-)

Comment (jhelie) [inline scipy/sparse/linalg/eigen/arpack/setup.py]: my bad, not sure how that happened in my editor. fixed.

**Review by perimosocordiae — APPROVED** [review]

**Comment by jhelie** [comment]
mmh travis failed due to a time out but was passing earlier...

**Comment by perimosocordiae** [comment]
Travis is happy, and so am I. Thanks @jhelie, merged.

**Comment by jhelie** [comment]
my pleasure @perimosocordiae . btw if you're interested here are your stats on lgtm.com: https://lgtm.com/people/13190038 :)

**Pull request merged**